File: solutions/erlang/collatz-conjecture/2/src/collatz_conjecture.erl
% SPDX-FileCopyrightText: 2025
%
% SPDX-License-Identifier: CC-BY-SA-4.0

-module(collatz_conjecture).

-export([steps/1]).

-spec steps(integer()) -> boolean().
steps(N) when N < 1 -> erlang:error(badarg);
steps(N) -> steps(N, 0).

steps(1, Steps) -> Steps;
steps(N, Steps) when N rem 2 == 0 -> steps(N div 2, Steps + 1);
steps(N, Steps) -> steps(3*N + 1, Steps + 1).

File: solutions/elixir/collatz-conjecture/4/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  @spec is_valid(any()) :: boolean()
  defguardp is_valid(value) when is_integer(value) and value > 0

  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(input), do: compute(input, 0)

  @spec compute(pos_integer(), non_neg_integer()) :: non_neg_integer()
  defp compute(1, steps), do: steps
  defp compute(number, steps) when is_valid(number) and rem(number, 2) === 0, do: compute(div(number, 2), steps + 1)
  defp compute(number, steps) when is_valid(number), do: compute((number*3)+1, steps + 1)
end
